Refactor - created BibliotecaAppTest and MainMenuInterface/Test

euloong opened the pull request:
- Created MainMenuInterface, MainMenuInterfaceTest and BibliotecaAppTest to simplify testing.
- Abstracted If-else condition out of checkOutBook into new method checkOutAvailableBook to provide correct message if book is available to be checked out or not.

Does this mean I can now replace the tests where I have been comparing the output strings with tests that call the required methods?

euloong commented:
- Renamed MainMenuInterfaceTest to DummyMainMenuInterface.
- Reduced tests in MainMenuTest to test only the strings created in MainMenu.
- Created tests in BibliotecaAppTest to test that methods in MainMenu have been called.
- Fixed where messages to the user were being displayed in the wrong sequence.

To Do:
- Rename MainMenuInterface to something sensible.
- Refactor `checkOutBook()` into smaller methods.

> - The top of checkOutAvailableBook looks like a duplicate of checkForAvailableBooks

I tried to combine them into one method but due to the order my messages are displayed I had to split them back into two methods.

> - I found out the other day that there is a Scanner.scanInt() method which will throw an exception automatically if a non-int is supplied... that would help clean up a little of checkOutBook

I tried to implement it but couldn't get it to work, I also previously tried using nextInt().

> - BTW, you can use this same approach to stub out the System.out and System.in integration as well: use an interface to abstract away the IO, create a test version which is easy to inject and query, and a real version which delegates to System.out and System.in.

I don't think I will have time to do this but I will try again after TWU.

andeemarks commented on test/com/twu/biblioteca/BibliotecaAppTest.java:
```diff
@@ -0,0 +1,62 @@
+public class BibliotecaAppTest {
+
+    @Test //[#1.2 & #1.3]
+    public void displayAvailableBooksCalled() {
+        DummyMainMenuInterface mainMenuInterface = new DummyMainMenuInterface();
+
+        mainMenuInterface.displayAvailableBooks();
+
+        assertTrue(mainMenuInterface.displayAvailableBooksHasBeenCalled);
+    }
```
I know this is a work-in-progress, so feel free to ignore this if it's going to change in subsequent commits, but in the test for `BibliotecaApp` you should naturally be creating an instance of `BibliotecaApp` and calling methods on that instance.  In this test, you are creating an instance of `DummyMainMenuInterface` and calling methods on that instance instead.  

The assertion is correct, but the test should be checking how that method has been called _indirectly_ via `BibliotecaApp`, rather than testing it directly.

Let me know if this does/doesn't make sense and I can jump on a quick call with you.

euloong replied:
I'll have to talk to you tomorrow. I tried to create an instance of BibliotecaApp in the test but intelliJ didn't like it due to the static methods used in BibliotecaApp.

euloong replied:
I tried to implement the following but not sure if I'm making the correct connections between the classes:

`@Test`
`public void displayAvailableBooksCalled() {`
       ` BibliotecaApp bibliotecaApp = new BibliotecaApp();`
        `bibliotecaApp.manageOptions(DummyMainMenu dummyMainMenu);`
       ` dummyMainMenu.displayAvailableBooks();`
        `assertTrue(dummyMainMenu.displayAvailableBooksHasBeenCalled);`
   ` }`

andeemarks replied:
That's _almost_ there!  You do not need to call `displayAvailableBooks` from your test as `BibliotecaApp` will be doing it for you when you ask it to `manageOptions`.  If you compare this to the sample one I sent through:

```
        Parent parent = new Parent();
        TestChild child = new TestChild();
        parent.doesStuff(child);

        assertTrue(child.hasBeenCalled);
```
* `parent` is the equivalent of your `bibliotecaApp`
* `child` is the equivalent of your `dummyMainMenu`
* `parent.doesStuff` is the equivalent of `bibliotecaApp.manageOptions`.

euloong replied:
I tried the below but the test runs and never completes, it just keeps running:

      
```
        BibliotecaApp bibliotecaApp = new BibliotecaApp();
        DummyMainMenuInterface dummyMainMenu = new DummyMainMenuInterface();
        bibliotecaApp.manageOptions(dummyMainMenu);

        assertTrue(dummyMainMenu.displayAvailableBooksHasBeenCalled);
```

andeemarks replied:
That is exactly how I would expect the test to look, so the fact the test never concludes suggests the loop is continuing to run... there are a couple of ways to address this:

1. Extract the switch statement into a method that only takes the user input and then call that method in your test (so you isolate the test from the loop), or
1. Have your test provide the keypresses to simulate the pressing of keys needed to exit the loop.

Looking through this stuff, I think Option 1 is an easier option to implement.

euloong commented:
- Renamed MainMenuInterface to DummyMainMenu.
- Refactored checkOutBook into smaller methods (finally worked out how to break it down).

euloong commented:
- Extracted switch statement into a new method switchOptions(String userInput, DummyMainMenu dummyMainMenu, boolean active) and updated the displayAvailableBooksCalled() test in BibliotecaAppTest.
- The test works and the app works except when I get to option 3 to quit the app. As I'm unable to link the active = false to the while(active) condition in manageOptions(), the app doesn't close on the user input.
I'll setup a call with you tomorrow to check what I'm doing wrong.

euloong commented:
- Extracted switch with the correct parameters. displayAvailableBooksCalled() test works.

euloong commented:
- Moved `active = true` from `switchOptions()` parameters
- Removed the else from `manageOptions()`
- Updated `BibliotecaApp` tests to focus only on the methods in `BibliotecaApp`
- Added tests for `MainMenuImplentation` back into `MainMenuImplementationTest`
